### ISSUE TITLE
inductor regression on aarch64 neoverse-v1, failing unit tests due to compiler error on gcc10.2

### ISSUE THREAD

**Issue by robert-hardwick** [pytorch/pytorch#137775]
### 🐛 Describe the bug

This is another neoverse-v1 / gcc optimization related issue. Very similar to #137597 but not the same issue.

I have checked and this error goes away when we add "-fno-tree-vectorize" compile flag.

```torch._inductor.exc.CppCompileError: C++ compile error
Command:
g++ /tmp/tmpuqk7lj9j/zx/czx2eyturb6j6m727xhvknkjbdu3y5nqqk66wgxcjkwnxuzvpm5r.cpp -D TORCH_INDUCTOR_CPP_WRAPPER -D C10_USING_CUSTOM_GENERATED_MACROS -D CPU_CAPABILITY_NEON -shared -fPIC -O3 -DNDEBUG -ffast-math -fno-finite-math-only -fno-unsafe-math-optimizations -ffp-contract=off -march=native -Wall -std=c++17 -Wno-unused-variable -Wno-unknown-pragmas -fopenmp -I/opt/_internal/cpython-3.10.14/include/python3.10 -I/opt/_internal/cpython-3.10.14/include/python3.10 -I/aarch64_env_test/lib/python3.10/site-packages/torch/include -I/aarch64_env_test/lib/python3.10/site-packages/torch/include/torch/csrc/api/include -I/aarch64_env_test/lib/python3.10/site-packages/torch/include/TH -I/aarch64_env_test/lib/python3.10/site-packages/torch/include/THC -I/aarch64_env_test/lib/python3.10/site-packages/torch/include -I/aarch64_env_test/lib/python3.10/site-packages/torch/include/torch/csrc/api/include -I/aarch64_env_test/lib/python3.10/site-packages/torch/include/TH -I/aarch64_env_test/lib/python3.10/site-packages/torch/include/THC -D_GLIBCXX_USE_CXX11_ABI=0 -ltorch -ltorch_cpu -ltorch_python -lc10 -lgomp -L/opt/_internal/cpython-3.10.14/lib -L/aarch64_env_test/lib/python3.10/site-packages/torch/lib -L/aarch64_env_test/lib/python3.10/site-packages/torch/lib -L/aarch64_env_test/lib/python3.10/site-packages/torch/lib -o /tmp/tmpuqk7lj9j/zx/czx2eyturb6j6m727xhvknkjbdu3y5nqqk66wgxcjkwnxuzvpm5r.so
Output:
during GIMPLE pass: slp
/tmp/tmpuqk7lj9j/zx/czx2eyturb6j6m727xhvknkjbdu3y5nqqk66wgxcjkwnxuzvpm5r.cpp: In function ‘void kernel(const float*, const float*, float*, int64_t*, int64_t, int64_t)’:
/tmp/tmpuqk7lj9j/zx/czx2eyturb6j6m727xhvknkjbdu3y5nqqk66wgxcjkwnxuzvpm5r.cpp:3:18: internal compiler error: in vect_get_vector_types_for_stmt, at tree-vect-stmts.c:12252
    3 | extern "C"  void kernel(const float* in_ptr0,
      |                  ^~~~~~
Please submit a full bug report,
with preprocessed source if appropriate.
```

To reproduce

On neoverse-v1 with gcc10.2

```
To execute this test, run the following from the base repo dir:
    python test/inductor/test_torchinductor_codegen_dynamic_shapes.py DynamicShapesCodegenCpuTests.test_fractional_max_pool2d3_dynamic_shapes_cpu
```

There are are a number of other tests that fail with the same error, but too many to include.

### Versions

Collecting environment information...
PyTorch version: 2.6.0.dev20241010+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (AltArch) (aarch64)
GCC version: (GCC) 10.2.1 20210130 (Red Hat 10.2.1-11)
Clang version: Could not collect
CMake version: version 3.30.4
Libc version: glibc-2.17

Python version: 3.10.15 (main, Sep 28 2024, 23:26:50) [GCC 10.2.1 20210130 (Red Hat 10.2.1-11)] (64-bit runtime)
Python platform: Linux-6.8.0-1015-aws-aarch64-with-glibc2.17
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:          aarch64
Byte Order:            Little Endian
CPU(s):                48
On-line CPU(s) list:   0-47
Thread(s) per core:    1
Core(s) per socket:    48
Socket(s):             1
NUMA node(s):          1
Model:                 1
BogoMIPS:              2100.00
L1d cache:             64K
L1i cache:             64K
L2 cache:              1024K
L3 cache:              32768K
NUMA node0 CPU(s):     0-47
Flags:                 fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma lrcpc dcpop sha3 sm3 sm4 asimddp sha512 sve asimdfhm dit uscat ilrcpc flagm ssbs paca pacg dcpodp svei8mm svebf16 i8mm bf16 dgh rng

Versions of relevant libraries:
[pip3] mypy==1.11.2
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.22.4
[pip3] onnx==1.16.1
[pip3] onnxscript==0.1.0.dev20240817
[pip3] optree==0.13.0
[pip3] torch==2.6.0.dev20241011+cpu

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @malfet @snadampal @milpuz01

**Comment by robert-hardwick** [comment]
@malfet another neoverse-v1 / gcc regression. This time, limited to gcc <= 10.

**Comment by robert-hardwick** [comment]
working on creating a small reproducer to raise bug with gcc

**Comment by malfet** [comment]
Hmm, should we just revert https://github.com/pytorch/pytorch/pull/134672

**Comment by robert-hardwick** [comment]
I don't know that #134672 is actually the cause, we don't get a very verbose stracktrace with this error. Again looks more like a gcc optimization problem on sve-enabled hardware.

I can manually revert this change and test locally.

As to why unit tests and benchmark runs are not seeing this: there are 2 reasons.

- The primary reason is that workflows are using gcc10 toolchain, the ci images are ( pytorch-linux-jammy-aarch64-py3.10-gcc11-inductor-benchmarks, and pytorch-linux-jammy-aarch64-py3.10-gcc11 ), both are gcc11 images. This bug is only present in gcc10.
- there seems to be disparity between unit tests and benchmark tests in terms of the runner type
    - unit tests seem to be running on t4g ( neoverse-n1 )
    - benchmark runner appears to be m7g ( neoverse-v1 )
    
unless i'm mistaken about the runner node types?

**Comment by robert-hardwick** [comment]
I'm actually testing this from inside a builder ( cpu-aarch64-main ) docker image which has gcc10.2 toolchain, on a neoverse-v1 machine. 

https://hub.docker.com/r/pytorch/manylinuxaarch64-builder/tags

**Comment by malfet** [comment]
How bad would be the SVE perf if compile without tree vectorization? Because a very simple solution seems to be adding `-fno-tree-vectorize` to https://github.com/pytorch/pytorch/blob/e30c55ee527b40d67555464b9e402b4b7ce03737/torch/_inductor/cpu_vec_isa.py#L173

**Comment by robert-hardwick** [comment]
I've attached the fix for this. I managed to identify the piece of code we need to prevent tree optimization on which i think will limit the performance impact. 

Also was seeing this issue before #134672 was merged.

I've not created the gcc bug report yet because i need to make a minimal reproducer, so I need to untangle all the pytorch specific code from it which will take some time. I've got a cpp-only reproducer that uses libtorch, but will need to trim it down before i can raise an issue with gcc.

I'll link it once i've done that.

**Comment by malfet** [comment]
@robert-hardwick thank you for applying those fixes, but do you know how much performance will be lost when those optimization are disabled. And perhaps it's better to just disable SVE for older compilers, as they could be a lot more problems like that...

**Comment by malfet** [comment]
> @robert-hardwick thank you for applying those fixes, but do you know how much performance will be lost when those optimization are disabled. And perhaps it's better to just disable SVE for older compilers, as they could be a lot more problems like that...

Also, I'm not sure if it makes much sense to report the bug to GCC as it was fixed in the later versions anyway.

**Comment by robert-hardwick** [comment]
Here is the generated code that torchinductor generates 
[torchcompileoutput.txt](https://github.com/user-attachments/files/17385302/torchcompileoutput.txt)

which i broke down into a reproducer for gcc
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=117161

It seems that fix in #137795 was somewhat accidental and i'm not confident that there are not other undiscovered issues. Therefore I think we should blanket disable GCC's tree optimizer with `-fno-tree-optimize` for GCC10 on neoverse-v1( -march=armv8-a+sve)

Possibly also `-fno-tree-loop-optimize`.

**Comment by robert-hardwick** [comment]
> @robert-hardwick thank you for applying those fixes, but do you know how much performance will be lost when those optimization are disabled. And perhaps it's better to just disable SVE for older compilers, as they could be a lot more problems like that...

If you see my comment above I think disabling gcc's auto vectorize optimizations for older versions of GCC would be a more targetted approach. 

`
if GCC_VERSION <= 10:
   _arch_flags +="-fno-tree-vectorize -fno-tree-loop-vectorize"
`

something like this... is there a way to get the GCC version from python?

**Comment by robert-hardwick** [comment]
@malfet looks like the issue has been marked as a duplicate and fixed in 10.3.0 - [https://gcc.gnu.org/bugzilla/show_bug.cgi?id=117161](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=117161)

r.e. my point above, do you think trying to read the gcc version from python is a good idea? or would you prefer just to blanket diable the optimizations on all gcc versions?
